File: src/0996-Number-of-Squareful-Arrays/0996.py
class Solution:
    def numSquarefulPerms(self, A: 'List[int]') -> 'int':
        n = len(A)
        res = 0
        def permute(nums, i):
            nonlocal res
            if i == n:
                res += 1
                return 
            for k in range(i, n):
                if i != k and nums[i] == nums[k]:
                    continue
                nums[i], nums[k] = nums[k], nums[i]
                if i == 0 or int((nums[i] + nums[i-1])**(0.5))**2 == (nums[i] + nums[i-1]):
                    permute(nums.copy(), i+1)
                
        A.sort()
        permute(A, 0)
        return res

File: src/0996-Number-of-Squareful-Arrays/0996.cpp
static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    int numSquarefulPerms(vector<int>& A) 
    {
        sort(A.begin(), A.end());
        int res = 0;
        permute(A, 0, res);
        return res;
    }
    
private:
    void permute(vector<int> A, int i, int& res) 
    {
        if (i == A.size()) 
        {
            ++res; return;
        }
        
        for (int j = i; j < A.size(); ++j) 
        {
            if (j > i && A[i] == A[j]) continue;
            
            swap(A[j], A[i]);
            
            if ((i == 0) or pow((int)sqrt(A[i] + A[i-1]), 2) == A[i] + A[i-1])
            {
                permute(A, i + 1, res);
            }
        }
    }
};